docs/Planner.md

```diff
@@ -1,63 +1,86 @@
+This document describes the design of the Flux query planner and optimizer.
 
-## Interface
+## Plans
+--------
 
-The Planner inter is defined as:
+A Flux query plan is represented as a Directed Acyclic Graph (DAG).
+The nodes of the graph represent individual query operations to perform.
+The edges of the graph represent the flow of data from one operation to another.
+In a Flux query plan, data flows vertically from the leaves of the DAG to the roots.
 
-```go
-type Planner interface {
-    Plan(Query, []Storage) Plan
-}
-```
+## Plan Nodes
+-------------
 
-The planner is responsible for taking a query DAG and a set of available storage interfaces and produce a plan DAG.
+```go
+type PlanNode interface {
+	// Returns an identifier for this plan node
+	ID() NodeID
 
-## Plans
+	// Returns the time bounds for this plan node
+	Bounds() *Bounds
 
-Plans are created via a two step process:
+	// Plan nodes executed immediately before this node
+	Predecessors() []PlanNode
 
-1. Create a general plan from the query.
-2. Create a specific plan from the general plan and available storage interface.
+	// Plan nodes executed immediately after this node
+	Successors() []PlanNode
 
-The general plan specifies all the needed data frames and their lineage needed to produce the final query result.
-The specific plan specifies how the general plan will be executed, which storage interfaces will be consumed and how.
+	// Specification of the procedure represented by this node
+	ProcedureSpec() ProcedureSpec
 
-The general plan does not leave the scope of the Planner and is not part of the API of the planner.
-Hence the Plan type above it the specific plan.
+	// Type of procedure represented by this node
+	Kind() ProcedureKind
 
-## Plan DAG
+    ...
+}
+```
 
-Both the general and specific plans are represented as DAGs.
-The nodes of the DAG represent data frames to be produced, while the edges of the DAG represent the operations need to construct the data frames.
-This is inverted from the Query DAG where the nodes are operations and edges represents data sets.
+The nodes of a query represent the operations to be performed.
+An operation represents a transformation on the data in some form.
+Nodes have predecessors and successors.
+A node's predecessors are the query operations executed immediately before it.
+A node's successors are the query operations executed immediately after it.
+A node receives data directly from its predecessors.
+A node feeds data directly to its successors.
 
-The leaves of the plan DAG represent sources of data and the data flows from bottom up through the tree.
-Again this is inverted from the Query DAG where data flows top down.
+## Planning
+-----------
 
-## Data Frames
+There are two types of query plans - logical and physical.
+The nodes of a logical plan represent logical operations.
+The nodes of a physical plan represent physical operations.
+Logical operations define how data will be transformed.
```

Comment:
I've heard it described this way: logical operations describe *what* happens to the data (join, sort, etc) and physical operations specify *how* the operations happen, i.e., the actual algorithms (merge join, bubble sort, etc).